types/link.go

```diff
@@ -0,0 +1,222 @@
+package types
+
+import (
+	"fmt"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+)
+
+// LinkCommonParams represents the common parameters for all link types.
+type LinkCommonParams struct {
+	Mtu    int                    `yaml:"mtu,omitempty"`
+	Labels map[string]string      `yaml:"labels,omitempty"`
+	Vars   map[string]interface{} `yaml:"vars,omitempty"`
+}
+
+// LinkDefinition represents a link definition in the topology file.
+type LinkDefinition struct {
+	Type string `yaml:"type"`
+	LinkConfig
+}
+
+// LinkDefinitionType represents the type of a link definition.
+type LinkDefinitionType string
+
+const (
+	LinkTypeVEth    LinkDefinitionType = "veth"
+	LinkTypeMgmtNet LinkDefinitionType = "mgmt-net"
+	LinkTypeMacVLan LinkDefinitionType = "macvlan"
+	LinkTypeMacVTap LinkDefinitionType = "macvtap"
+	LinkTypeHost    LinkDefinitionType = "host"
+
+	// LinkTypeBrief is a link definition where link types
+	// are encoded in the endpoint definition as string and allow users
+	// to quickly type out link endpoints in a yaml file.
+	LinkTypeBrief LinkDefinitionType = "brief"
+)
+
+// parseLinkType parses a string representation of a link type into a LinkDefinitionType.
+func parseLinkType(s string) (LinkDefinitionType, error) {
+	switch strings.TrimSpace(strings.ToLower(s)) {
+	case string(LinkTypeMacVLan):
+		return LinkTypeMacVLan, nil
+	case string(LinkTypeMacVTap):
+		return LinkTypeMacVTap, nil
+	case string(LinkTypeVEth):
+		return LinkTypeVEth, nil
+	case string(LinkTypeMgmtNet):
+		return LinkTypeMgmtNet, nil
+	case string(LinkTypeHost):
+		return LinkTypeHost, nil
+	case string(LinkTypeBrief):
+		return LinkTypeBrief, nil
+	default:
+		return "", fmt.Errorf("unable to parse %q as LinkType", s)
+	}
+}
+
+var _ yaml.Unmarshaler = (*LinkDefinition)(nil)
+var _ yaml.Marshaler = (*LinkDefinition)(nil)
+
+// MarshalYAML serializes LinkDefinition (e.g when used with generate command).
+// As of now it falls back to converting the LinkConfig into a
+// RawVEthLink, such that the generated LinkConfigs adhere to the new LinkDefinition
+// format instead of the brief one.
+func (r *LinkDefinition) MarshalYAML() (interface{}, error) {
+	rawVEth, err := vEthFromLinkConfig(&r.LinkConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	x := struct {
+		RawVEthLink `yaml:",inline"`
+		Type        string `yaml:"type"`
+	}{
+		RawVEthLink: *rawVEth,
+		Type:        string(LinkTypeVEth),
+	}
+
+	return x, nil
+}
```

Comment:
@steiler I wonder, why do we currently need a custom marshaller, since `(*CLab).Links` contains LinkDefinitions that are not different from the old/current implementation?

Reply:
If you ask we, I do not want to get rid of LinkDefinition in the internal code. It should be the Raw input format (see comment below) and be converted to the specific type with resolved string references (e.g. string id -> node reference).

Yes I did think of allowing the old / brief format only for the veth links which I assume is at least 90% of the cases. And force the new format onto people for all the other link types.
If you think this is the wrong path... fine, we can broaden the accepted brief definition types.

Reply:
in [b8e4ecd](https://github.com/srl-labs/containerlab/pull/1450/commits/b8e4ecda23049259e91b9b435baeafe2045bd541) I removed the custom unmarshaller for LinkDefinition, and made `generate` command to use brief format by default.

This removed a few funcs as a consequence.

Reply:
it seems you removed the marshaller not the unmarshaller

Reply:
Yeah, I meant to say marshaller 😅